core/shared/src/main/scala/zio/FiberPool.scala

```diff
@@ -0,0 +1,184 @@
+package zio
+
+import scala.collection.immutable.{ Queue => IQueue }
+
+import FiberPool.{ Shutdown, State }
+
+/**
+ * A fiber-based equivalent of an ExecutorService. A fiber pool lazily spawns fibers that
+ * handle submitted tasks, and keeps them alive for additional tasks submitted.
+ *
+ * Tasks can be submitted using the [[FiberPool#submit]] method; the pool may be shutdown
+ * using the [[FiberPool#shutdown]] method.
+ *
+ * Fiber pools are usually more efficient for long-running processes in which tasks
+ * must be executed in a bounded amount of parallelism compared to repeatedly spawning
+ * new fibers.
+ */
+class FiberPool(state: Ref[State], workerLimit: Long, defectHandler: Cause[Nothing] => UIO[Any]) {
+
+  /**
+   * Submits a new task to the pool. This method will semantically block until a fiber is available
+   * to start the task. Defects caused by the task will be handled by the pool's defect handler.
+   *
+   * If the pool is already shutdown, this method will interrupt its caller.
+   */
+  def submit[R](task: URIO[R, Any]): URIO[R, Any] =
+    ZIO.environment[R].flatMap { r =>
+      Promise.make[Nothing, Unit].flatMap { taskReceived =>
```

Comment:
Good point to specify fiber id when you make the promise?

Reply:
This is the same situation where we don't know which fiber is going to resolve it, because it's going to be resolved by one of the fibers in the pool. 🤷🏽‍♂️